GLM confounds as numpy array are not supported

ok something does not make sense about allowing numpy array: they are accepted as inputs but never appended to the output list... so they will never be passed on to whatever needs them later...

I think that in a follow up PR I will check if we actually support array as confounds and it is properly implemented or if this is a doc string error.

_Originally posted by @Remi-Gau in https://github.com/nilearn/nilearn/pull/4807#discussion_r1869224785_

The following test fails: confounds cannot be passed as numpy arrays to first level

```python
import pandas as pd 

from nilearn.glm.first_level import FirstLevelModel

from nilearn._utils.data_gen import (
    basic_paradigm,
    write_fake_fmri_data_and_design,
)

def test_fmri_inputs_with_confounds_numpy(tmp_path):
    """Test with confounds and, events or design matrix."""
    n_timepoints = 10
    shapes = ((3, 4, 5, n_timepoints),)
    mask, func_img, _ = write_fake_fmri_data_and_design(
        shapes, file_path=tmp_path
    )

    conf = pd.DataFrame([0] * n_timepoints, columns=["conf"])

    events = basic_paradigm()

    func_img = func_img[0]

    flm = FirstLevelModel(mask_img=mask, t_r=2.0).fit(
        func_img, confounds=conf.values, events=events,
    )
```

```
nilearn/glm/first_level/first_level.py:867: in fit
    self.design_matrices_ = self._create_all_designs(
        bins       = 100
        confounds  = []
        design_matrices = None
        events     = [  trial_type  onset  duration
0         c0     30       1.0
1         c0     70       1.0
2         c0    100       1.0
3         c1     10       1.0
4         c1     30       1.0
5         c1     90       1.0
6         c2     30       1.0
7         c2     40       1.0
8         c2     60       1.0]
        run_imgs   = ['/tmp/pytest-of-remi/pytest-47/test_fmri_inputs_with_confound0/fmri_run0.nii']
        sample_masks = None
        self       = FirstLevelModel(fir_delays=[0],
                mask_img=PosixPath('/tmp/pytest-of-remi/pytest-47/test_fmri_inputs_with_confound0/mask.nii'),
                memory=Memory(location=None), t_r=2.0)
nilearn/glm/first_level/first_level.py:633: in _create_all_designs
    design = self._create_single_design(
        confounds  = []
        design_matrices = []
        events     = [  trial_type  onset  duration
0         c0     30       1.0
1         c0     70       1.0
2         c0    100       1.0
3         c1     10       1.0
4         c1     30       1.0
5         c1     90       1.0
6         c2     30       1.0
7         c2     40       1.0
8         c2     60       1.0]
        n_scans    = 10
        run_idx    = 0
        run_img    = <nibabel.nifti1.Nifti1Image object at 0x7119f20c4110>
        run_imgs   = ['/tmp/pytest-of-remi/pytest-47/test_fmri_inputs_with_confound0/fmri_run0.nii']
        self       = FirstLevelModel(fir_delays=[0],
                mask_img=PosixPath('/tmp/pytest-of-remi/pytest-47/test_fmri_inputs_with_confound0/mask.nii'),
                memory=Memory(location=None), t_r=2.0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = FirstLevelModel(fir_delays=[0],
                mask_img=PosixPath('/tmp/pytest-of-remi/pytest-47/test_fmri_inputs_with_confound0/mask.nii'),
                memory=Memory(location=None), t_r=2.0)
n_scans = 10
events = [  trial_type  onset  duration
0         c0     30       1.0
1         c0     70       1.0
2         c0    100       1...         c1     90       1.0
6         c2     30       1.0
7         c2     40       1.0
8         c2     60       1.0]
confounds = [], run_idx = 0

    def _create_single_design(self, n_scans, events, confounds, run_idx):
        """Build experimental design of a single run."""
        if confounds is not None:
>           confounds_matrix = confounds[run_idx].to_numpy()
E           IndexError: list index out of range

confounds  = []
events     = [  trial_type  onset  duration
0         c0     30       1.0
1         c0     70       1.0
2         c0    100       1.0
3         c1     10       1.0
4         c1     30       1.0
5         c1     90       1.0
6         c2     30       1.0
7         c2     40       1.0
8         c2     60       1.0]
n_scans    = 10
run_idx    = 0
self       = FirstLevelModel(fir_delays=[0],
                mask_img=PosixPath('/tmp/pytest-of-remi/pytest-47/test_fmri_inputs_with_confound0/mask.nii'),
                memory=Memory(location=None), t_r=2.0)

nilearn/glm/first_level/first_level.py:644: IndexError
```

--- a/doc/changes/latest.rst
+++ b/doc/changes/latest.rst
@@ -18,6 +18,8 @@ Fixes
 
 - :bdg-dark:`Code` Fix all occurrences of the RUF012 error related to mutable default values in Python classes (:gh:`4954` by `Idrissa Traore`_).
 
+- :bdg-dark:`Code` Support for confounds as numpy arrays in :meth:`nilearn.glm.first_level.FirstLevelModel.fit` (:gh:`4967` by `Rémi Gau`_).
+
 Enhancements
 ------------
 

--- a/nilearn/_utils/glm.py
+++ b/nilearn/_utils/glm.py
@@ -13,6 +13,8 @@ def check_and_load_tables(tables_to_check, var_name):
        if they are pandas.DataFrame, \
        or a CSV or TSV file that can be loaded to a pandas.DataFrame.
 
+       Numpy arrays will also be appended as is.
+
     tables_to_check : str or pathlib.Path to a TSV or CSV \
               or pandas.DataFrame or numpy.ndarray or, \
               a list of str or pathlib.Path to a TSV or CSV \
@@ -27,7 +29,7 @@ def check_and_load_tables(tables_to_check, var_name):
 
     Returns
     -------
-    list of pandas.DataFrame
+    list of pandas.DataFrame or numpy.arrays
 
     Raises
     ------
@@ -55,7 +57,7 @@ def check_and_load_tables(tables_to_check, var_name):
         if isinstance(table, str):
             loaded = _read_events_table(table)
             tables.append(loaded)
-        elif isinstance(table, pd.DataFrame):
+        elif isinstance(table, (pd.DataFrame, np.ndarray)):
             tables.append(table)
 
     return tables

--- a/nilearn/glm/first_level/first_level.py
+++ b/nilearn/glm/first_level/first_level.py
@@ -639,19 +639,38 @@ class FirstLevelModel(BaseGLM):
         return design_matrices
 
     def _create_single_design(self, n_scans, events, confounds, run_idx):
-        """Build experimental design of a single run."""
+        """Build experimental design of a single run.
+
+        Parameters
+        ----------
+        n_scans: int
+
+        events : list of pandas.DataFrame
+
+        confounds : list of pandas.DataFrame or numpy.arrays
+
+        run_idx : int
+        """
+        confounds_matrix = None
+        confounds_names = None
         if confounds is not None:
-            confounds_matrix = confounds[run_idx].to_numpy()
+            confounds_matrix = confounds[run_idx]
+
+            if isinstance(confounds_matrix, pd.DataFrame):
+                confounds_names = confounds[run_idx].columns.tolist()
+                confounds_matrix = confounds_matrix.to_numpy()
+            else:
+                # create dummy names when dealing with numpy arrays
+                confounds_names = [
+                    f"confound_{i}" for i in range(confounds_matrix.shape[1])
+                ]
+
             if confounds_matrix.shape[0] != n_scans:
                 raise ValueError(
                     "Rows in confounds does not match "
                     "n_scans in run_img "
                     f"at index {run_idx}."
                 )
-            confounds_names = confounds[run_idx].columns.tolist()
-        else:
-            confounds_matrix = None
-            confounds_names = None
 
         start_time = self.slice_time_ref * self.t_r
         end_time = (n_scans - 1 + self.slice_time_ref) * self.t_r
@@ -773,7 +792,8 @@ class FirstLevelModel(BaseGLM):
             Each column in a DataFrame corresponds to a confound variable
             to be included in the regression model of the respective run_img.
             The number of rows must match the number of volumes in the
-            respective run_img. Ignored in case designs is not None.
+            respective run_img.
+            Ignored in case designs is not None.
             If string, then a path to a csv file is expected.
 
         sample_masks : array_like, or :obj:`list` of array_like, default=None

--- a/nilearn/glm/tests/test_first_level.py
+++ b/nilearn/glm/tests/test_first_level.py
@@ -656,31 +656,76 @@ def test_fmri_inputs_events_type(tmp_path):
 
 
 def test_fmri_inputs_with_confounds(tmp_path):
-    """Test with confounds and, events or design matrix."""
+    """Test with confounds and, events."""
+    n_timepoints = 10
+    shapes = ((3, 4, 5, n_timepoints),)
+    mask, func_img, _ = write_fake_fmri_data_and_design(
+        shapes, file_path=tmp_path
+    )
+
+    conf = pd.DataFrame([0] * n_timepoints, columns=["conf"])
+
+    events = basic_paradigm()
+
+    func_img = func_img[0]
+
+    # Provide t_r, confounds, and events but no design matrix
+    flm = FirstLevelModel(mask_img=mask, t_r=2.0).fit(
+        func_img,
+        confounds=conf,
+        events=events,
+    )
+    assert "conf" in flm.design_matrices_[0]
+
+    # list are OK
+    FirstLevelModel(mask_img=mask, t_r=2.0).fit(
+        func_img,
+        confounds=[conf],
+        events=events,
+    )
+
+    # test with confounds as numpy array
+    flm = FirstLevelModel(mask_img=mask, t_r=2.0).fit(
+        func_img,
+        confounds=conf.to_numpy(),
+        events=events,
+    )
+    assert "confound_0" in flm.design_matrices_[0]
+
+    flm = FirstLevelModel(mask_img=mask, t_r=2.0).fit(
+        func_img,
+        confounds=[conf.to_numpy()],
+        events=events,
+    )
+    assert "confound_0" in flm.design_matrices_[0]
+
+
+def test_fmri_inputs_confounds_ingored_with_design_matrix(tmp_path):
+    """Test with confounds with design matrix.
+
+    Confounds ignored if design matrix is passed
+    """
     n_timepoints = 10
     shapes = ((3, 4, 5, n_timepoints),)
     mask, func_img, des = write_fake_fmri_data_and_design(
         shapes, file_path=tmp_path
     )
 
-    conf = pd.DataFrame([0, 0])
-
-    events = basic_paradigm()
+    conf = pd.DataFrame([0] * n_timepoints, columns=["conf"])
 
     func_img = func_img[0]
-    des = des[0]
 
-    # Provide t_r, confounds, and events but no design matrix
-    FirstLevelModel(mask_img=mask, t_r=2.0).fit(
-        func_img,
-        confounds=pd.DataFrame([0] * n_timepoints, columns=["conf"]),
-        events=events,
+    des = pd.read_csv(
+        des[0],
+        sep="\t",
+    )
+    n_col_in_des = len(des.columns)
+
+    flm = FirstLevelModel(mask_img=mask).fit(
+        func_img, confounds=conf, design_matrices=des
     )
 
-    # test with confounds as numpy array
-    FirstLevelModel(mask_img=mask).fit(
-        [func_img], confounds=conf.values, design_matrices=[des]
-    )
+    assert len(flm.design_matrices_[0].columns) == n_col_in_des
 
 
 def test_fmri_inputs_errors(tmp_path, shape_4d_default):

--- a/nilearn/glm/tests/test_first_level.py
+++ b/nilearn/glm/tests/test_first_level.py
@@ -700,7 +700,7 @@ def test_fmri_inputs_with_confounds(tmp_path):
     assert "confound_0" in flm.design_matrices_[0]
 
 
-def test_fmri_inputs_confounds_ingored_with_design_matrix(tmp_path):
+def test_fmri_inputs_confounds_ignored_with_design_matrix(tmp_path):
     """Test with confounds with design matrix.
 
     Confounds ignored if design matrix is passed

--- a/nilearn/glm/tests/test_first_level.py
+++ b/nilearn/glm/tests/test_first_level.py
@@ -715,10 +715,7 @@ def test_fmri_inputs_confounds_ignored_with_design_matrix(tmp_path):
 
     func_img = func_img[0]
 
-    des = pd.read_csv(
-        des[0],
-        sep="\t",
-    )
+    des = pd.read_csv(des[0], sep="\t")
     n_col_in_des = len(des.columns)
 
     flm = FirstLevelModel(mask_img=mask).fit(

